src/inmanta/data/__init__.py

```diff
@@ -6634,59 +6592,65 @@ async def set_last_processed_model_version(
 ]
 
 
-def set_connection_pool(pool: asyncpg.pool.Pool) -> None:
-    LOGGER.debug("Connecting data classes")
-    for cls in _classes:
-        cls.set_connection_pool(pool)
-
-
-async def disconnect() -> None:
-    LOGGER.debug("Disconnecting data classes")
-    # Enable `return_exceptions` to make sure we wait until all close_connection_pool() calls are finished
-    # or until the gather itself is cancelled.
-    result = await asyncio.gather(*[cls.close_connection_pool() for cls in _classes], return_exceptions=True)
-    exceptions = [r for r in result if r is not None and isinstance(r, Exception)]
-    if exceptions:
-        raise exceptions[0]
-
-
 PACKAGE_WITH_UPDATE_FILES = inmanta.db.versions
 
 # Name of core schema in the DB schema verions
 # prevent import loop
 CORE_SCHEMA_NAME = schema.CORE_SCHEMA_NAME
 
 
-async def connect(
-    host: str,
-    port: int,
-    database: str,
-    username: str,
-    password: str,
-    create_db_schema: bool = True,
-    connection_pool_min_size: int = 10,
-    connection_pool_max_size: int = 10,
-    connection_timeout: float = 60,
-) -> asyncpg.pool.Pool:
-    pool = await asyncpg.create_pool(
-        host=host,
-        port=port,
-        database=database,
-        user=username,
-        password=password,
-        min_size=connection_pool_min_size,
-        max_size=connection_pool_max_size,
-        timeout=connection_timeout,
-    )
+async def stop_engine():
+    global ENGINE
+    if ENGINE is not None:
+        await ENGINE.dispose()
+
+
+async def start_engine(
+    url: str,
+    pool_size: int = 10,
+    max_overflow: int = 0,
+    pool_timeout: float = 60.0,
+    echo: bool = False,
+):
+    """
+    engine vs connection vs session overview
+    https://stackoverflow.com/questions/34322471/sqlalchemy-engine-connection-and-session-difference
+    """
+    global ENGINE
+
+    if ENGINE is not None:
+        if not inmanta.RUNNING_TESTS:
+            raise Exception("Engine already running: cannot call start_engine twice.")
+        await ENGINE.dispose(close=True)
+
+    LOGGER.debug("Creating engine...")
     try:
-        set_connection_pool(pool)
-        if create_db_schema:
-            async with pool.acquire() as con:
-                await schema.DBSchema(CORE_SCHEMA_NAME, PACKAGE_WITH_UPDATE_FILES, con).ensure_db_schema()
```

Comment:
Where did the schema manager go?

Reply:
It was lifted in [databaseservice.initialize_sql_alchemy_engine](https://github.com/inmanta/inmanta-core/blob/af2640281546f6c858d298567ed9810704e8e7bd/src/inmanta/server/services/databaseservice.py#L290)